src/NHibernate/Transform/AliasToBeanResultTransformer.cs

```
@@ -179,10 +120,93 @@ private bool TrySet(string alias, object value, object resultObj, Dictionary<str
+				: null;
+			for (int i = 0; i < aliases.Length; i++)
+			{
+				string alias = aliases[i];
+				if (string.IsNullOrEmpty(alias))
+					continue;
+
+				var memberInfo = GetMemberInfo(alias);
+				var valueExpr = Expression.ArrayAccess(tupleParam, Expression.Constant(i));
+				var valueToAssign = GetTyped(memberInfo, valueExpr, getException);
+				if (wrapInVariables)
+				{
+					var variable = Expression.Variable(valueToAssign.Type);
+					variableAndAssigmentDic[variable] = Expression.Assign(variable, valueToAssign);
+					valueToAssign = variable;
+				}
+				bindings.Add(Expression.Bind(memberInfo, valueToAssign));
+			}
+			Expression initExpr = Expression.MemberInit(Expression.New(_resultClass), bindings);
+			if (!ResultClass.IsClass)
+				initExpr = Expression.Convert(initExpr, typeof(object));
+
+			if (wrapInVariables)
+			{
+				initExpr = Expression.Block(variableAndAssigmentDic.Keys, variableAndAssigmentDic.Values.Concat(new[] { initExpr }));
+			}
+			return (Func<object[], object>) Expression.Lambda(initExpr, tupleParam).Compile();
+		}
+
+		private bool ShouldWrapInVariables()
+		{
+#if NETFX
+			//On .net461 throws if DTO is struct: TryExpression is not supported as a child expression when accessing a member on type '[TYPE]' because it is a value type.
+			if (!_resultClass.IsClass)
+				return true;
+#endif
+			return false;
+		}
+
+		private static Expression GetTyped(MemberInfo memberInfo, Expression expr, Expression<Func<string, object, Exception>> getEx)
+		{
+			var type = GetMemberType(memberInfo);
+			if (type == typeof(object))
+				return expr;
+
+			var originalValue = expr;
+			if (!type.IsClass)
+			{
+				expr = Expression.Coalesce(expr, Expression.Default(type));
+			}
+
+			return Expression.TryCatch(
+					Expression.Convert(expr, type),
+					Expression.Catch(
+						typeof(InvalidCastException),
+						Expression.Throw(Expression.Invoke(getEx, Expression.Constant(memberInfo.ToString()), originalValue), type)
```

The catched exception should be transmitted as inner exception to the raised one. More "User friendly" exceptions are not so user friendly when they cause a loss of information by swallowing the original exception.

---

But there is really nothing useful in codegenerated "IvalidCastException" exception. Empty message and no useful stacktrace.